File: Antidote/TabBarBadgeItem.m
//
//  TabBarBadgeItem.m
//  Antidote
//
//  Created by Dmytro Vorobiov on 09.07.15.
//  Copyright (c) 2015 dvor. All rights reserved.
//

#import <BlocksKit/UIControl+BlocksKit.h>
#import <Masonry/Masonry.h>

#import "TabBarBadgeItem.h"
#import "AppearanceManager.h"

static const CGFloat kImageAndTextContainerYOffset = 2.0;
static const CGFloat kImageAndTextOffset = 3.0;

static const CGFloat kBadgeTopOffset = -5.0;
static const CGFloat kBadgeHorizontalOffset = 5.0;
static const CGFloat kBadgeMinimumWidth = 22.0;
static const CGFloat kBadgeHeight = 18.0;

@interface TabBarBadgeItem ()

@property (strong, nonatomic) UIView *imageAndTextContainer;
@property (strong, nonatomic) UIImageView *imageView;
@property (strong, nonatomic) UILabel *textLabel;

@property (strong, nonatomic) UIView *badgeContainer;
@property (strong, nonatomic) UILabel *badgeLabel;

@property (strong, nonatomic) UIButton *button;

@end

@implementation TabBarBadgeItem
@synthesize selected = _selected;
@synthesize didTapOnItem = _didTapOnItem;

#pragma mark -  Lifecycle

- (instancetype)init
{
    self = [super init];

    if (! self) {
        return nil;
    }

    self.backgroundColor = [UIColor clearColor];

    [self createImageAndTextViews];
    [self createBadgeViews];
    [self createButton];

    [self installConstraints];

    return self;
}

- (void)layoutSubviews
{
    [super layoutSubviews];

    self.badgeContainer.layer.cornerRadius = self.badgeContainer.frame.size.height / 2;
}

#pragma mark -  Properties

- (void)setImage:(UIImage *)image
{
    self.imageView.image = [image imageWithRenderingMode:UIImageRenderingModeAlwaysTemplate];
}

- (UIImage *)image
{
    return self.imageView.image;
}

- (void)setText:(NSString *)text
{
    self.textLabel.text = text;
}

- (NSString *)text
{
    return self.textLabel.text;
}

- (void)setBadgeText:(NSString *)badgeText
{
    self.badgeLabel.text = badgeText;
    self.badgeContainer.hidden = (badgeText.length == 0);
}

- (NSString *)badgeText
{
    return self.badgeLabel.text;
}

#pragma mark -  TabBarItemProtocol

- (void)setSelected:(BOOL)selected
{
    _selected = selected;

    UIColor *color = selected ? [[AppContext sharedContext].appearance textMainColor] : [UIColor colorWithWhite:0.5 alpha:1.0];

    self.textLabel.textColor = color;
    self.imageView.tintColor = color;
}

#pragma mark -  Private

- (void)createImageAndTextViews
{
    self.imageAndTextContainer = [UIView new];
    self.imageAndTextContainer.backgroundColor = [UIColor clearColor];
    [self addSubview:self.imageAndTextContainer];

    self.imageView = [UIImageView new];
    self.imageView.backgroundColor = [UIColor clearColor];
    [self.imageAndTextContainer addSubview:self.imageView];

    self.textLabel = [UILabel new];
    self.textLabel.backgroundColor = [UIColor clearColor];
    self.textLabel.font = [[AppContext sharedContext].appearance fontHelveticaNeueWithSize:10.0];
    [self.imageAndTextContainer addSubview:self.textLabel];
}

- (void)createBadgeViews
{
    self.badgeContainer = [UIView new];
    self.badgeContainer.backgroundColor = [[AppContext sharedContext].appearance statusBusyColor];
    self.badgeContainer.layer.masksToBounds = YES;
    [self addSubview:self.badgeContainer];

    self.badgeLabel = [UILabel new];
    self.badgeLabel.textColor = [UIColor whiteColor];
    self.badgeLabel.textAlignment = NSTextAlignmentCenter;
    self.badgeLabel.backgroundColor = [UIColor clearColor];
    self.badgeLabel.font = [[AppContext sharedContext].appearance fontHelveticaNeueWithSize:14.0];
    [self.badgeContainer addSubview:self.badgeLabel];

    // hide badge
    self.badgeText = nil;
}

- (void)createButton
{
    self.button = [UIButton new];
    self.button.backgroundColor = [UIColor clearColor];
    [self addSubview:self.button];

    weakself;
    [self.button bk_addEventHandler:^(UIButton *b) {
        strongself;
        if (self.didTapOnItem) {
            self.didTapOnItem(self);
        }
    } forControlEvents:UIControlEventTouchUpInside];
}

- (void)installConstraints
{
    [self.imageAndTextContainer makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.equalTo(self);
        make.centerY.equalTo(self).offset(kImageAndTextContainerYOffset);
    }];

    [self.imageView makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.imageAndTextContainer);
        make.centerX.equalTo(self.imageAndTextContainer);
    }];

    [self.textLabel makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.imageView.bottom).offset(kImageAndTextOffset);
        make.centerX.equalTo(self.imageAndTextContainer);
        make.bottom.equalTo(self.imageAndTextContainer);
    }];

    [self.badgeContainer makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.imageAndTextContainer.left);
        make.top.equalTo(self.imageAndTextContainer.top).offset(kBadgeTopOffset);
        make.width.greaterThanOrEqualTo(kBadgeMinimumWidth);
        make.height.equalTo(kBadgeHeight);
    }];

    [self.badgeLabel makeConstraints:^(MASConstraintMaker *make) {
        make.left.equalTo(self.badgeContainer).offset(kBadgeHorizontalOffset);
        make.right.equalTo(self.badgeContainer).offset(-kBadgeHorizontalOffset);
        make.centerY.equalTo(self.badgeContainer);
    }];

    [self.button makeConstraints:^(MASConstraintMaker *make) {
        make.edges.equalTo(self);
    }];
}

@end

File: Antidote/TabBarBadgeItem.swift
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

import UIKit
import SnapKit

private struct Constants {
    static let ImageAndTextContainerYOffset = 2.0
    static let ImageAndTextOffset = 3.0

    static let BadgeTopOffset = -5.0
    static let BadgeHorizontalOffset = 5.0
    static let BadgeMinimumWidth = 22.0
    static let BadgeHeight: CGFloat = 18.0
}

class TabBarBadgeItem: TabBarAbstractItem {
    override var selected: Bool {
        didSet {
            let color = theme.colorForType(selected ? .TabItemActive : .TabItemInactive)

            textLabel.textColor = color
            imageView.tintColor = color
        }
    }

    var image: UIImage? {
        didSet {
            imageView.image = image?.withRenderingMode(.alwaysTemplate)
        }
    }
    var text: String? {
        didSet {
            textLabel.text = text
        }
    }

    var badgeText: String? {
        didSet {
            badgeTextWasUpdated()
        }
    }

    /// If there is any badge text, accessibilityValue will be set to <badgeText + badgeAccessibilityEnding>
    var badgeAccessibilityEnding: String?

    fileprivate let theme: Theme

    fileprivate var imageAndTextContainer: UIView!
    fileprivate var imageView: UIImageView!
    fileprivate var textLabel: UILabel!

    fileprivate var badgeContainer: UIView!
    fileprivate var badgeLabel: UILabel!

    fileprivate var button: UIButton!

    init(theme: Theme) {
        self.theme = theme

        super.init(frame: CGRect.zero)

        backgroundColor = .clear

        createViews()
        installConstraints()

        badgeTextWasUpdated()
    }

    required convenience init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

// Accessibility
extension TabBarBadgeItem {
    override var accessibilityLabel: String? {
        get {
            return text
        }
        set {}
    }

    override var accessibilityValue: String? {
        get {
            guard var result = badgeText else {
                return nil
            }

            if let ending = badgeAccessibilityEnding {
                result += " " + ending
            }

            return result
        }
        set {}
    }
}

// Actions
extension TabBarBadgeItem {
    func buttonPressed() {
        didTapHandler?()
    }
}

private extension TabBarBadgeItem {
    func createViews() {
        imageAndTextContainer = UIView()
        imageAndTextContainer.backgroundColor = .clear
        addSubview(imageAndTextContainer)

        imageView = UIImageView()
        imageView.backgroundColor = .clear
        imageAndTextContainer.addSubview(imageView)

        textLabel = UILabel()
        textLabel.textColor = theme.colorForType(.NormalText)
        textLabel.textAlignment = .center
        textLabel.backgroundColor = .clear
        textLabel.font = UIFont.systemFont(ofSize: 10.0)
        imageAndTextContainer.addSubview(textLabel)

        badgeContainer = UIView()
        badgeContainer.backgroundColor = theme.colorForType(.TabBadgeBackground)
        badgeContainer.layer.masksToBounds = true
        badgeContainer.layer.cornerRadius = Constants.BadgeHeight / 2
        addSubview(badgeContainer)

        badgeLabel = UILabel()
        badgeLabel.textColor = theme.colorForType(.TabBadgeText)
        badgeLabel.textAlignment = .center
        badgeLabel.backgroundColor = .clear
        badgeLabel.font = UIFont.antidoteFontWithSize(14.0, weight: .light)
        badgeContainer.addSubview(badgeLabel)

        button = UIButton()
        button.backgroundColor = .clear
        button.addTarget(self, action: #selector(TabBarBadgeItem.buttonPressed), for: .touchUpInside)
        addSubview(button)
    }

    func installConstraints() {
        imageAndTextContainer.snp.makeConstraints {
            $0.centerX.equalTo(self)
            $0.centerY.equalTo(self).offset(Constants.ImageAndTextContainerYOffset)
        }

        imageView.snp.makeConstraints {
            $0.top.equalTo(imageAndTextContainer)
            $0.centerX.equalTo(imageAndTextContainer)
        }

        textLabel.snp.makeConstraints {
            $0.top.equalTo(imageView.snp.bottom).offset(Constants.ImageAndTextOffset)
            $0.centerX.equalTo(imageAndTextContainer)
            $0.bottom.equalTo(imageAndTextContainer)
        }

        badgeContainer.snp.makeConstraints {
            $0.leading.equalTo(imageAndTextContainer.snp.leading)
            $0.top.equalTo(imageAndTextContainer.snp.top).offset(Constants.BadgeTopOffset)
            $0.width.greaterThanOrEqualTo(Constants.BadgeMinimumWidth)
            $0.height.equalTo(Constants.BadgeHeight)
        }

        badgeLabel.snp.makeConstraints {
            $0.leading.equalTo(badgeContainer).offset(Constants.BadgeHorizontalOffset)
            $0.trailing.equalTo(badgeContainer).offset(-Constants.BadgeHorizontalOffset)
            $0.centerY.equalTo(badgeContainer)
        }

        button.snp.makeConstraints {
            $0.edges.equalTo(self)
        }
    }

    func badgeTextWasUpdated() {
        badgeLabel.text = badgeText
        badgeContainer.isHidden = (badgeText == nil) || badgeText!.isEmpty
    }
}
